dump2tar/man/dump2tar.1
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.\" Macro for inserting an option description prologue.
.\" .OD <long> [<short>] [args]
.de OD
.  ds args "
.  if !'\\$3'' .as args \fI\\$3\fP
.  if !'\\$4'' .as args \\$4
.  if !'\\$5'' .as args \fI\\$5\fP
.  if !'\\$6'' .as args \\$6
.  if !'\\$7'' .as args \fI\\$7\fP
.  PD 0
.  if !'\\$2'' .IP "\fB\-\\$2\fP \\*[args]" 4
.  if !'\\$1'' .IP "\fB\-\-\\$1\fP \\*[args]" 4
.  PD
..
.\" Macro for inserting code line.
.\" .CL <text>
.de CL
.  ds pfont \\n[.f]
.  nh
.  na
.  ft CR
\\$*
.  ft \\*[pfont]
.  ad
.  hy
.  br
..
.\" Macro for inserting a man page reference.
.\" .MP man-page section [suffix]
.de MP
.  nh
.  na
.  BR \\$1 (\\$2)\\$3
.  ad
.  hy
..
.
.TH "dump2tar" "1" "2016\-09\-02" "" ""
.
.SH "NAME"
dump2tar - Gather file contents and command output into a tar archive
.
.
.SH "SYNOPSIS"
.B "dump2tar "
.RI "[" "OPTIONS" "] " "SPECS"
.
.
.SH "DESCRIPTION"
.B dump2tar
creates a tar archive from the contents of any files, including files of
unknown size.

Examples for files of unknown size are:
.IP \(bu 3
Named pipes (FIFOs)
.PP
.IP \(bu 3
Particular Linux kernel debugfs or sysfs files
.PP
.IP \(bu 3
Character or block devices
.PP

When adding such a file,
.B dump2tar
first reads all available data until an end-of-file indication is found. From
this data, it then creates a regular file entry in the resulting tar archive.
By default, symbolic links and directories are preserved in the archive in
their original form.

.B dump2tar
can also:
.IP \(bu 3
Add files under a different name
.PP
.IP \(bu 3
Run arbitrary commands and add the resulting command output as a
regular file
.PP
.
.
.SH "FILE SPECIFICATIONS"
.
This section describes the format of the
.I SPECS
argument mentioned in the command synopsis.
Use the following command line syntax to identify data sources and
to specify file names within the archive:
.PP

.TP
.I "PATH"
Adds the contents of the file system subtree at file system location
.I PATH
(with possible exceptions described by options) in the archive under the same
file name as on the file system.
.PP
.
.
.TP
.IR "FILENAME" ":=" "PATH"
Adds the contents of the file at file system location
.I PATH
in the archive under the name specified by
.IR FILENAME .
.PP
.
.
.TP
.IR "FILENAME" "|=" "CMDLINE"
Runs the command
.IR CMDLINE
and captures both the resulting standard output and standard error streams.
Adds the collected output as a regular file named
.I FILENAME
in the resulting archive. You can also include the resulting program exit code
by using option \-\-add\-cmd\-status.
.PP
.
You can also specify "\-\-". All specifications that follow are interpreted as
simple file names. This is useful for archiving files that contain ":=" or "|=".
.PP
.
.
.SH "OUTPUT OPTIONS"
.
.OD "output\-file" "o" "TARFILE"
Writes the resulting tar archive to
.IR TARFILE .
An existing file at the specified file system location is overwritten.

If this option is omitted or if "\-" is specified for
.IR TARFILE ,
the archive is written to the standard output stream.
.PP
.
.
.OD "gzip" "z" ""
Compresses the resulting tar archive using gzip.
.PP
.
.
.OD "max\-size" "m" "VALUE"
Sets an upper size limit, in bytes, for the resulting archive. If this limit
is exceeded after adding a file, no further files are added.
.PP
.
.
.OD "timeout" "t" "VALUE"
Sets an upper time limit, in seconds, for the archiving process. If this limit
is exceeded while adding a file, that file is truncated and no
further files are added.
.PP
.
.
.OD "no-eof" "" ""
Does not write an end-of-file marker.

Use this option if you want to create an archive that can be extended by
appending additional tar archive data.

Note: Do not use this option for the final data to be added.
A valid tar archive requires a trailing end-of-file marker.
.PP
.
.
.OD "append" "" ""
Appends data to the end of the archive.

Use this option to incrementally build a tar file by repeatedly calling
.BR dump2tar .
You must specify the \-\-no\-eof option for each but the final call of
.BR dump2tar .
.PP
.
.
.OD "add-cmd-status" "" ""
Adds a separate file named
.RI \(dq FILENAME .cmdstatus\(dq
for each command output added through the
.RI \(dq FILENAME |= CMDLINE \(dq
notation (see FILE SPECIFICATIONS).
This file contains information about the exit status of the
process that executed the command:
.
.RS 8
.TP
.RI EXITSTATUS= VALUE
Unless
.I VALUE
is -1, the process ended normally with the specified exit value.
.PP
.
.TP
.RI TERMSIG= VALUE
Unless
.I VALUE
is -1, the process was stopped by a signal of the specified number.
.PP
.
.TP
.RI WAITPID_ERRNO= VALUE
Unless
.I VALUE
is -1, an attempt to obtain the status of the process failed with the
specified error.
.PP
.RE
.
.
.
.SH "INPUT OPTIONS"
.
.OD "files\-from" "F" "FILENAME"
Reads input data specifications (see FILE SPECIFICATIONS) from
.IR FILENAME ,
one specification per line. Each line contains either a file name or a
.IR FILENAME := PATH
or
.IR FILENAME |= CMDLINE
specification. Empty lines are ignored.

A line can also consist of only "\-\-". All lines following this specification
are interpreted as simple file names. This is useful for archiving files that
contain ":=" or "|=".
.PP
.
.
.OD "ignore\-failed\-read" "i" ""
Continues after read errors.

By default,
.B dump2tar
stops processing after encountering errors while reading an input file.
With this option,
.B dump2tar
prints a warning message and adds an empty entry for the erroneous file in
the archive.
.PP
.
.
.OD "buffer\-size" "b" "VALUE"
Reads data from input files in chunks of
.I VALUE
bytes. Large values can accelerate the archiving process for large files
at the cost of increased memory usage. The default value is 1048576.
.PP
.
.
.OD "file\-timeout" "T" "VALUE"
Sets an upper time limit, in seconds, for reading an input file.

.B dump2tar
stops processing a file when the time limit is exceeded. Archive entries for
such files are truncated to the amount of data that is collected by the time
the limit is reached.
.PP
.
.
.OD "file\-max\-size" "M" "N"
Sets an upper size limit, in bytes, for an input file.

.B dump2tar
stops processing a file when the size limit is exceeded. Archive entries for
such files are truncated to the specified size.
.PP
.
.
.OD "jobs" "j" "N"
By default,
.B dump2tar
processes one file at a time. With this option,
.B dump2tar
processes
.I N
files in parallel.

Parallel processing can accelerate the archiving process,
especially if input files are located on slow devices, or when output from
multiple commands is added to the archive.

Note: Use
.B tar
option \-\-delay\-directory\-restore when extracting files from an archive
created with \-\-jobs to prevent conflicts with directory permissions and
modification times.
.PP
.
.
.OD "jobs\-per\-cpu" "J" "N"
Processes
.I N
files for each online CPU in parallel.

Parallel processing can accelerate the
archiving process, especially if input files are located on slow devices, or
when output from multiple commands is added to the archive.

Note: Use
.B tar
option \-\-delay\-directory\-restore when extracting files from an archive
created with \-\-jobs\-per\-cpu to prevent conflicts with directory permissions
and modification times.
.PP
.
.
.OD "exclude" "x" "PATTERN"
Does not add files to the archive if their file names match
.IR PATTERN .
.I PATTERN
is an expression that uses the shell wildcards.
.PP
.
.
.OD "exclude\-from" "X" "FILENAME"
Does not add files to the archive if their names match at least one of the
patterns listed in the pattern file with name
.IR FILENAME .
In the pattern file, each line specifies an expression that uses the
shell wildcards.
.PP
.
.
.OD "exclude\-type" "" "TYPE"
Does not add files to the archive if they match at least one of the file types
specified with
.IR TYPE .
.I TYPE
uses one or more of the characters "fdcbpls", where:

.RS 8
.IP f 3
regular files
.PP
.IP d 3
directories
.PP
.IP c 3
character devices
.PP
.IP b 3
block devices
.PP
.IP p 3
named pipes (FIFOs)
.PP
.IP l 3
symbolic links
.PP
.IP s 3
sockets
.PP
.RE
.
.PP
.
.
.OD "dereference" "" ""
Adds the content of link targets instead of symbolic links.
.PP
.
.
.OD "no\-recursion" "" ""
Does not add files from sub\-directories.

By default,
.B dump2tar
adds archive entries for specified directories, and for the files within these
directories. With this option, a specified directory results in a single entry
for the directory. Any contained files to be included must be specified
explicitly.
.PP
.
.
.SH "MISC OPTIONS"
.
.OD "help" "h" ""
Prints an overview of available options, then exits.
.PP
.
.
.OD "verbose" "V" ""
Prints additional informational output.
.PP
.
.
.OD "quiet" "q" ""
Suppresses printing of informational output.
.PP
.
.
.
.SH "EXAMPLES"
.
.\fB
.CL # dump2tar a b \-o archive.tar
.\fR

.RS 4
Creates a tar archive named archive.tar containing files a and b.
.RE
.PP
.
.\fB
.CL # dump2tar /proc \-o procdump.tar.gz \-z \-i \-T 1 \-M 1048576
.\fR

.RS 4
Creates a gzip compressed tar archive named procdump.tar.gz that contains
all procfs files. Unreadable files are ignored. Files are truncated when the
first of the two limiting conditions is reached, either 1048576 bytes of
content or the reading time of 1 second.
.RE
.PP
.
.\fB
.CL # dump2tar '|=dmesg' '|=lspci' \-o data.tar
.\fR

.RS 4
Creates a tar archive named data.tar containing the output of the 'dmesg'
and 'lspci' commands.
.RE
.PP
.
.\fB
.CL # dump2tar /sys/kernel/debug/ -x '*/tracing/*' -o debug.tar -i
.\fR

.RS 4
Creates a tar archive named debug.tar containing the contents of directory
/sys/kernel/debug/ while excluding any file that is located in a sub-directory
named 'tracing'.
.RE
.PP
.
.
.SH "EXIT CODES"
.TP
.B 0
The program finished successfully
.TP
.B 1
A run-time error occurred
.TP
.B 2
The specified command was not valid
.PP
.
.
.SH "SEE ALSO"
.MP dump2tar 1 ,
.MP tar 1


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option
  - lshwc: Add support for delta counter value display
  - lspai: Add output '--format' option
  - lsreipl: Add secure boot state to output
  - lswhc: Add short names to lshwc output
  - pv_tools: Add Bash and Zsh completions
  - pvapconfig: Add '--unbind' option
  - pvimg/boot: Print error messages from stage3a bootloader
  - pvimg: Add support for CCK update
  - pvsecret: Add support for CCK update
  - pvsecret: Allow retrieving secrets by index & warn for duplicated entries
  - pvsecret: Deny adding secrets with duplicated secret IDs
  - zdev: Add support for virtio devices
  - zipl: Enhance mirror support
  - zipl: Implement '--dry-run' option for all dump jobs
  - zipl_helper.device-mapper: Support mirrors over NVMe devices
  - zkey/dracut: Add a dracut config file for zkey
  - zkey/initramfs: Update initramfs hook to correct drivers and include zkey plugins
  - zkey: Add support for converting a clear-key LUKS2 volume to use a secure key

  Bug Fixes:
  - chpstat: Add missing CMG 5 data fields
  - chpstat: Fix DPU utilization calculation
  - libutil/util_file: Handle over-read in util_file_read_fd()
  - pvattest: Fix successful 'check' evaluation
  - pvsecret: Fix some edge cases for plaintext keys
  - zipl_helper.device-mapper: Fix imprecise is_device_mapper() predicate
  - zkey: Fix EP11 secure key reencipher function
  - zpcictl: Fix command line parsing for invalid options

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution

* __v2.36.0 (2024-12-06)__

  For Linux kernel version: 6.12

  s390-tools: Define Rust MSRV as 1.75.0

  Add new tools / libraries:
  - cpacfinfo: Tool to provide CPACF information
  - opticsmon: Tools to monitor optical modules for directly attached PCI based NICs
  - pvimg: Rust rewrite of genprotimg

  Changes of existing tools:
  - chpstat: Add data bandwidth utilization column
  - chpstat: Add support for full CMCB
  - chpstat: Add support for new CMG types
  - dbginfo.sh: add overview commands and crypto update
  - genprotimg:
    1. genprotimg is now a symbolic link to the new tool `pvimg create`
    2. Breaking API changes in genprotimg:
       1. An existing output file is no longer silently overwritten; to revert
          to the old behavior use the `--overwrite` option.
       2. The Linux kernel component is now checked to determine whether it is a
          binary s390x kernel. To disable this behavior use the
          `--no-component-check` option.
  - hyptop: Support for structured output (json, json-seq, csv)
  - lszfcp: Add missing fallback marker for non-good fc_host port_state
  - lszfcp: Improve speed with many SCSI devices
  - pvattest: Add attestation policy check command
  - zipl: Add support of partitions of mirror md-devices

  Bug Fixes:
  - lszcrypt: Fix wrong state showing up for removed AP queue within SE guest

dump2tar/src/dump2tar.c
/*
 * dump2tar - tool to dump files and command output into a tar archive
 *
 * Command line interface
 *
 * Copyright IBM Corp. 2016, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "lib/util_opt.h"
#include "lib/util_prg.h"

#include "dump.h"
#include "global.h"
#include "idcache.h"
#include "misc.h"
#include "strarray.h"

#define MIN_BUFFER_SIZE		4096

#define	OPT_NOSHORT_BASE	256

#define	OPT_DEREFERENCE		(OPT_NOSHORT_BASE + 0)
#define OPT_NORECURSION		(OPT_NOSHORT_BASE + 1)
#define OPT_EXCLUDETYPE		(OPT_NOSHORT_BASE + 2)

/* Program description */
static const struct util_prg dump2tar_prg = {
	.desc = "Use dump2tar to create a tar archive from the contents "
		"of arbitrary files.\nIt works even when the size of actual "
		"file content is not known beforehand,\nsuch as with FIFOs, "
		"character devices or certain Linux debugfs or sysfs files.\n"
		"\nYou can also add files under different names and add "
		"command output using the\nformat described in section SPECS "
		"below. When no additional options are\nspecified, the "
		"resulting archive is written to the standard output stream\n"
		"in uncompressed tar format.",
	.args = "SPECS",
	.copyright_vec = {
		{ "IBM Corp.", 2016, 2017 },
		UTIL_PRG_COPYRIGHT_END
	},
};

/* Definition of command line options */
static struct util_opt dump2tar_opts[] = {
	UTIL_OPT_SECTION("OUTPUT OPTIONS"),
	{
		.option = { "output-file", required_argument, NULL, 'o' },
		.argument = "FILE",
		.desc = "Write archive to FILE (default: standard output)",
	},
#ifdef HAVE_ZLIB
	{
		.option = { "gzip", no_argument, NULL, 'z' },
		.desc = "Write a gzip compressed archive",
	},
#endif /* HAVE_ZLIB */
	{
		.option = { "max-size", required_argument, NULL, 'm' },
		.argument = "N",
		.desc = "Stop adding files when archive size exceeds N bytes",
	},
	{
		.option = { "timeout", required_argument, NULL, 't' },
		.argument = "SEC",
		.desc = "Stop adding files after SEC seconds",
	},
	{
		.option = { "no-eof", no_argument, NULL, 131 },
		.desc = "Do not write an end-of-file marker",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "add-cmd-status", no_argument, NULL, 132 },
		.desc = "Add status of commands as separate file",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "append", no_argument, NULL, 133 },
		.desc = "Append output to end of file",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},

	UTIL_OPT_SECTION("INPUT OPTIONS"),
	{
		.option = { "files-from", required_argument, NULL, 'F' },
		.argument = "FILE",
		.desc = "Read filenames from FILE (- for standard input)",
	},
	{
		.option = { "ignore-failed-read", no_argument, NULL, 'i' },
		.desc = "Continue after read errors",
	},
	{
		.option = { "buffer-size", required_argument, NULL, 'b' },
		.argument = "N",
		.desc = "Read data in chunks of N byte (default: 16384)",
	},
	{
		.option = { "file-timeout", required_argument, NULL, 'T' },
		.desc = "Stop reading file after SEC seconds",
		.argument = "SEC",
	},
	{
		.option = { "file-max-size", required_argument, NULL, 'M' },
		.argument = "N",
		.desc = "Stop reading file after N bytes",
	},
	{
		.option = { "jobs", required_argument, NULL, 'j' },
		.argument = "N",
		.desc = "Read N files in parallel (default: 1)",
	},
	{
		.option = { "jobs-per-cpu", required_argument, NULL, 'J' },
		.argument = "N",
		.desc = "Read N files per CPU in parallel",
	},
	{
		.option = { "exclude", required_argument, NULL, 'x' },
		.argument = "PATTERN",
		.desc = "Don't add files matching PATTERN",
	},
	{
		.option = { "exclude-from", required_argument, NULL, 'X' },
		.argument = "FILE",
		.desc = "Don't add files matching patterns in FILE",
	},
	{
		.option = { "exclude-type", required_argument, NULL,
			    OPT_EXCLUDETYPE },
		.argument = "TYPE",
		.desc = "Don't add files of specified TYPE (one of: fdcbpls)",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "dereference", no_argument, NULL, OPT_DEREFERENCE },
		.desc = "Add link targets instead of links",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "no-recursion", no_argument, NULL,
			    OPT_NORECURSION },
		.desc = "Don't add files from sub-directories",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},

	UTIL_OPT_SECTION("MISC OPTIONS"),
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	{
		.option = { "verbose", no_argument, NULL, 'V' },
		.desc = "Print additional informational output",
	},
	{
		.option = { "quiet", no_argument, NULL, 'q' },
		.desc = "Suppress printing of informational output",
	},
	UTIL_OPT_END,
};

/* Split buffer size specification in @arg into two numbers to be stored in
 * @from_ptr and @to_ptr. Return %EXIT_OK on success. */
static int parse_buffer_size(char *arg, size_t *from_ptr, size_t *to_ptr)
{
	char *err;
	unsigned long from, to;

	if (!*arg) {
		mwarnx("Empty buffer size specified");
		return EXIT_USAGE;
	}

	from = strtoul(arg, &err, 10);

	if (*err == '-')
		to = strtoul(err + 1, &err, 10);
	else
		to = *to_ptr;

	if (*err) {
		mwarnx("Invalid buffer size: %s", arg);
		return EXIT_USAGE;
	}

	if (from < MIN_BUFFER_SIZE || to < MIN_BUFFER_SIZE) {
		mwarnx("Buffer size too low (minimum %u)", MIN_BUFFER_SIZE);
		return EXIT_USAGE;
	}

	if (to < from)
		to = from;

	*from_ptr = from;
	*to_ptr = to;

	return EXIT_OK;
}

static void parse_and_add_spec(struct dump_opts *opts, const char *spec)
{
	char *op, *s, *inname, *outname = NULL;
	bool is_cmd = false;

	s = mstrdup(spec);
	op = strstr(s, "|=");
	if (op)
		is_cmd = true;
	else
		op = strstr(s, ":=");

	if (op) {
		*op = 0;
		inname = op + 2;
		outname = s;
	} else {
		inname = s;
	}
	dump_opts_add_spec(opts, inname, outname, is_cmd);
	free(s);
}

static int add_specs_from_file(struct dump_opts *opts, const char *filename)
{
	FILE *fd;
	char *line = NULL;
	size_t line_size;
	int rc = EXIT_RUNTIME;
	bool need_close = false, parse_spec = true;

	if (strcmp(filename, "-") == 0)
		fd = stdin;
	else {
		fd = fopen(filename, "r");
		if (!fd) {
			mwarn("%s: Cannot open file", filename);
			goto out;
		}
		need_close = true;
	}

	while ((getline(&line, &line_size, fd) != -1)) {
		chomp(line, "\n");
		if (line[0] == 0)
			continue;
		if (parse_spec && strcmp(line, "--") == 0) {
			/* After a line containing --, no more := or |= specs
			 * are expected */
			parse_spec = false;
			continue;
		}
		if (parse_spec)
			parse_and_add_spec(opts, line);
		else
			dump_opts_add_spec(opts, line, NULL, false);
	}

	if (ferror(fd))
		mwarn("%s: Cannot read file", filename);
	else
		rc = EXIT_OK;

out:
	if (need_close)
		fclose(fd);
	free(line);

	return rc;
}

static void print_help(void)
{
	static const struct {
		const char *name;
		const char *desc;
	} specs[] = {
		{ "PATH", "Add file or directory at PATH" },
		{ "NEWPATH:=PATH", "Add file or directory at PATH as NEWPATH" },
		{ "NEWPATH|=CMDLINE", "Add output of command line CMDLINE as "
		  "NEWPATH" },
		{ NULL, NULL },
	};
	int i;

	util_prg_print_help();
	printf("SPECS\n");
	for (i = 0; specs[i].name; i++)
		util_opt_print_indented(specs[i].name, specs[i].desc);
	printf("\n");
	util_opt_print_help();
}

int main(int argc, char *argv[])
{
	int rc = EXIT_USAGE, opt;
	long i;
	struct dump_opts *opts;

	if (getenv("DUMP2TAR_DEBUG"))
		global_debug = true;

	util_prg_init(&dump2tar_prg);
	util_opt_init(dump2tar_opts, "-");
	misc_init();

	opts = dump_opts_new();
	opterr = 0;
	while ((opt = util_opt_getopt_long(argc, argv)) != -1) {
		switch (opt) {
		case 'h': /* --help */
			print_help();
			rc = EXIT_OK;
			goto out;
		case 'v': /* --version */
			util_prg_print_version();
			rc = EXIT_OK;
			goto out;
		case 'V': /* --verbose */
			global_verbose = true;
			global_quiet = false;
			opts->verbose = true;
			opts->quiet = false;
			break;
		case 'q': /* --quiet */
			global_quiet = true;
			global_verbose = false;
			opts->quiet = true;
			opts->verbose = false;
			break;
		case 'i': /* --ignore-failed-read */
			opts->ignore_failed_read = true;
			break;
		case 'j': /* --jobs N */
			opts->jobs = atoi(optarg);
			if (opts->jobs < 1) {
				mwarnx("Invalid number of jobs: %s", optarg);
				goto out;
			}
			break;
		case 'J': /* --jobs-per-cpu N */
			opts->jobs_per_cpu = atoi(optarg);
			if (opts->jobs_per_cpu < 1) {
				mwarnx("Invalid number of jobs: %s", optarg);
				goto out;
			}
			break;
		case 'b': /* --buffer-size N */
			if (parse_buffer_size(optarg, &opts->read_chunk_size,
					      &opts->max_buffer_size))
				goto out;
			break;
		case 'x': /* --exclude PATTERN */
			add_str_to_strarray(&opts->exclude, optarg);
			break;
		case 'X': /* --exclude-from FILE */
			if (add_file_to_strarray(&opts->exclude, optarg))
				goto out;
			break;
		case 'F': /* --files-from FILE */
			if (add_specs_from_file(opts, optarg))
				goto out;
			break;
		case 'o': /* --output-file FILE */
			if (opts->output_file) {
				mwarnx("Output file specified multiple times");
				goto out;
			}
			opts->output_file = optarg;
			break;
		case OPT_DEREFERENCE: /* --dereference */
			opts->dereference = true;
			break;
		case OPT_NORECURSION: /* --no-recursion */
			opts->recursive = false;
			break;
		case OPT_EXCLUDETYPE: /* --exclude-type TYPE */
			for (i = 0; optarg[i]; i++) {
				if (dump_opts_set_type_excluded(opts,
								optarg[i]))
					break;

			}
			if (optarg[i]) {
				mwarnx("Unrecognized file type: %c", optarg[i]);
				goto out;
			}
			break;
		case 131: /* --no-eof */
			opts->no_eof = true;
			break;
		case 132: /* --add-cmd-status */
			opts->add_cmd_status = true;
			break;
		case 133: /* --append */
			opts->append = true;
			break;
		case 't': /* --timeout VALUE */
			opts->timeout = atoi(optarg);
			if (opts->timeout < 1) {
				mwarnx("Invalid timeout value: %s", optarg);
				goto out;
			}
			break;
		case 'T': /* --file-timeout VALUE */
			opts->file_timeout = atoi(optarg);
			if (opts->file_timeout < 1) {
				mwarnx("Invalid timeout value: %s", optarg);
				goto out;
			}
			break;
		case 'm': /* --max-size N */
			opts->max_size = atol(optarg);
			if (opts->max_size < 2) {
				mwarnx("Invalid maximum size: %s", optarg);
				goto out;
			}
			break;
		case 'M': /* --file-max-size N */
			opts->file_max_size = atol(optarg);
			if (opts->file_max_size < 2) {
				mwarnx("Invalid maximum size: %s", optarg);
				goto out;
			}
			break;
		case 'z': /* --gzip */
			opts->gzip = true;
			break;
		case 1: /* Filename specification or unrecognized option */
			if (optarg[0] == '-') {
				mwarnx("Invalid option '%s'", optarg);
				goto out;
			}
			parse_and_add_spec(opts, optarg);
			break;
		case '?': /* Unrecognized option */
			if (optopt)
				mwarnx("Invalid option '-%c'", optopt);
			else
				mwarnx("Invalid option '%s'", argv[optind - 1]);
			goto out;
		case ':': /* Missing argument */
			mwarnx("Option '%s' requires an argument",
			      argv[optind - 1]);
			goto out;
		default:
			break;
		}
	}
	if (optind >= argc && opts->num_specs == 0) {
		mwarnx("Please specify files to dump");
		goto out;
	}

	for (i = optind; i < argc; i++)
		dump_opts_add_spec(opts, argv[i], NULL, false);

	rc = dump_to_tar(opts);

out:
	idcache_cleanup();
	misc_cleanup();
	dump_opts_free(opts);

	if (rc == EXIT_USAGE)
		util_prg_print_parse_error();

	return rc;
}
